src/query/functions/utils/heap.go

```diff
@@ -0,0 +1,200 @@
+// THE SOFTWARE.
+
+package utils
+
+import (
+	"container/heap"
+)
+
+// ValueIndexPair is a pair of float value and index at which it exists
+type ValueIndexPair struct {
+	Val   float64
+	Index int
+}
+
+type lessFn func(ValueIndexPair, ValueIndexPair) bool
+
+func maxHeapLess(i, j ValueIndexPair) bool {
+	if i.Val == j.Val {
+		return i.Index > j.Index
+	}
+	return i.Val < j.Val
+}
+
+func minHeapLess(i, j ValueIndexPair) bool {
+	if i.Val == j.Val {
+		return i.Index > j.Index
+	}
+	return i.Val > j.Val
+}
+
+// FloatHeap is a heap that can be given a maximum size
+type FloatHeap struct {
+	isMaxHeap bool
+	capacity  int
+	floatHeap *floatHeap
+}
+
+// NewFloatHeap builds a new FloatHeap based on first parameter
+// and a capacity given by second parameter. Zero and negative
+// values for maxSize provide an unbounded FloatHeap
+func NewFloatHeap(isMaxHeap bool, capacity int) FloatHeap {
+	var less lessFn
+	if isMaxHeap {
+		less = maxHeapLess
+	} else {
+		less = minHeapLess
+	}
+
+	floatHeap := &floatHeap{
+		heap: make([]ValueIndexPair, 0, capacity),
+		less: less,
+	}
+
+	heap.Init(floatHeap)
+	return FloatHeap{
+		isMaxHeap: isMaxHeap,
+		capacity:  capacity,
+		floatHeap: floatHeap,
+	}
+}
+
+// Push pushes a value and index pair to the heap
+func (fh FloatHeap) Push(value float64, index int) {
+	h := fh.floatHeap
+	// If capacity is zero or negative, allow infinite size heap
+	if fh.capacity > 0 {
+		// At max size, drop or replace incoming value.
+		// Otherwise, continue as normal
+		if len(h.heap) >= fh.capacity {
+			peek := h.heap[0]
+			// Compare incoming value with current top of heap.
+			// Decide if to replace the current top, or to drop incoming
+			// value as appropriate for this min/max heap
+			//
+			// If values are equal, do not add incoming value regardless
+			//
+			// NB(arnikola): unfortunately, can't just replace first
+			// element as it may not respect internal order. Need to
+			// run heap.Fix() to rectify this
+			if fh.isMaxHeap && value > peek.Val ||
+				(!fh.isMaxHeap && value < peek.Val) {
+				h.heap[0] = ValueIndexPair{
+					Val:   value,
+					Index: index,
+				}
+
+				heap.Fix(h, 0)
+			}
+
+			return
+		}
+
+		// Otherwise, fallthrough
+	}
+
+	heap.Push(h, ValueIndexPair{
+		Val:   value,
+		Index: index,
+	})
+}
+
+// Len returns the current length of the heap
+func (fh FloatHeap) Len() int {
+	return fh.floatHeap.Len()
+}
+
+// Cap returns the capacity of the heap
+func (fh FloatHeap) Cap() int {
+	return fh.capacity
+}
+
+// Reset resets the heap
+func (fh *FloatHeap) Reset() {
+	fh.floatHeap.heap = fh.floatHeap.heap[:0]
+}
+
+// Flush flushes the float heap and resets it. Does not guarantee order.
+func (fh FloatHeap) Flush() []ValueIndexPair {
+	values := fh.floatHeap.heap
+	fh.Reset()
+	return values
+}
+
+// FlushByPop flushes the float heap and resets it by popping values from the heap.
+// Guarantees order.
+func (fh FloatHeap) FlushByPop() []ValueIndexPair {
```

Comment:
is this only for used in tests ? then don't export it ?

Reply:
Might just delete this; was initially a bit reticent about pulling the underlying array as is done in `Flush`, but considering that the value index is included in the output data, it's unnecessary to have the output be sorted.